components/layout/query.rs

```diff
@@ -927,44 +931,48 @@ pub fn process_element_inner_text_query<N: LayoutNode>(node: N,
     inner_text.into_iter().collect()
 }
 
-// https://html.spec.whatwg.org/multipage/#inner-text-collection-steps
 #[allow(unsafe_code)]
-fn inner_text_collection_steps<N: LayoutNode>(node: N,
-                                              indexable_text: &IndexableText,
-                                              results: &mut Vec<InnerTextItem>) {
-    let mut items = Vec::new();
-    for child in node.traverse_preorder() {
-        let node = match child.type_id() {
-            LayoutNodeType::Text => {
-                child.parent_node().unwrap()
-            },
-            _ => child,
-        };
+fn get_style<N: LayoutNode>(node: &N) -> Option<ServoArc<ComputedValues>> {
+    let node = match node.type_id() {
+        LayoutNodeType::Text => {
+            node.parent_node().unwrap()
+        },
+        _ => *node,
+    };
 
-        let element_data = unsafe {
-            node.get_style_and_layout_data().map(|d| {
-                &(*(d.ptr.as_ptr() as *mut StyleData)).element_data
-            })
-        };
+    let element_data = unsafe {
+        node.get_style_and_layout_data().map(|d| {
+            &(*(d.ptr.as_ptr() as *mut StyleData)).element_data
+        })
+    };
 
-        if element_data.is_none() {
-            continue;
-        }
+    element_data.map_or(None, |element_data| {
+        element_data.borrow().styles.get_primary().map_or(None, |style| {
+            Some(style.clone())
+        })
+    })
+}
 
-        let style = match element_data.unwrap().borrow().styles.get_primary() {
-            None => continue,
-            Some(style) => style.clone(),
+// https://html.spec.whatwg.org/multipage/#inner-text-collection-steps
+fn inner_text_collection_steps<N: LayoutNode>(mut iter: TreeIterator<N>,
+                                              indexable_text: &IndexableText,
+                                              results: &mut Vec<InnerTextItem>) {
+    if let Some(child) = iter.next() {
+        let mut items = Vec::new();
+        let style = match get_style(&child) {
+            None => return,
+            Some(style) => style,
         };
 
         // Step 2.
         if style.get_inheritedbox().visibility != Visibility::Visible {
-            continue;
+            return;
         }
 
         // Step 3.
         let display = style.get_box().display;
         if !child.is_in_document() || display == Display::None {
```

Comment:
How could the child not be in the document but have style? You should be able to debug_assert it (I know it's pre-existing).